File: code/cvc5mvn/src/main/java/plaid/logic/GenEntailVerifier.java
package plaid.logic;

import io.github.cvc5.CVC5ApiException;
import io.github.cvc5.Sort;
import io.github.cvc5.Term;
import io.github.cvc5.TermManager;

import plaid.ast.*;
import plaid.cvc.TermFactory;
import plaid.cvc.Verifier;
import plaid.eval.ConstraintChecker;

import java.util.*;

public class GenEntailVerifier {
    private int counter;
    private final TermManager termManager;
    private final Sort sort;

    private final Program program;
    
    public GenEntailVerifier(Program program, String order) throws CVC5ApiException {
        counter = 1;
        termManager  = new TermManager();
        sort = termManager.mkFiniteFieldSort(order, 10);
        this.program = program;
    }
    
    private Str genFreshString(){
        return new Str("$" + counter++);
    }

    private Num genFreshCID() {
        return new Num(-counter++);
    }

    public Expr genFreshValue(Type type) {
        // TODO I need to update to use _ for unnamed variables
        return switch (type) {
            case StringType x -> genFreshString();
            case PartyIndexType x -> genFreshCID();
            case RecordType t -> {
                TreeMap<Identifier, Expr> map = new java.util.TreeMap<>(Map.of());
                for (Map.Entry<Identifier, Type> element : t.elements().entrySet()){
                    map.put(element.getKey(), genFreshValue(element.getValue()));
                }
                yield new FieldExpr(map);
            }
            default -> throw new RuntimeException("Unsupported type " + type);
        };
    }
    

    /**
     * generalized rule for entailment, verify if precondition entails postcondition for all inputs
     * @param e1 abstract preposition
     * @param e2 abstract postcondition
     * @return true/false
     */
    public boolean genEntails(List<TypedIdentifier> typings, Constraint e1, Constraint e2) {
        Map<Identifier, Expr> bindingList = new HashMap<>();
        
        // generate fresh values for variables 
        for (TypedIdentifier typing: typings){
            bindingList.put(typing.y(), genFreshValue(typing.t()));
        }
        
        // evalConstraint abstractConstraints into ground constraint
        ConstraintEvaluator evaluator = new ConstraintEvaluator(program);
        evaluator.binding_list.add(bindingList);
        Constraint groundE1 = evaluator.evalConstraint(e1);
        Constraint groundE2 =  evaluator.evalConstraint(e2);
        evaluator.binding_list.removeLast();

        // check if groundE1 and groundE2 are in fact ground
        if(ConstraintChecker.checkConstraint(groundE1) && ConstraintChecker.checkConstraint(groundE2)){
            // check if ground constraint1 entails ground constraint2 using Verifier entails
            TermFactory termFactory = new TermFactory(termManager, sort);
            Term groundE1Term = termFactory.constraintToTerm(groundE1);
            Term groundE2Term = termFactory.constraintToTerm(groundE2);
            Verifier verifier = new Verifier(termFactory);

            return verifier.entails(groundE1Term, groundE2Term);
        }
        else {
            throw new RuntimeException("constraints with fresh values are not ground!");
        }
    }

}

File: code/cvc5mvn/src/main/scala/plaid/logic/GenEntailVerifier.scala
package plaid.logic

import io.github.cvc5.{Sort, TermManager}
import plaid.ast.*
import plaid.cvc.{TermFactory, Verifier}
import plaid.eval.ConstraintChecker

import java.util.{HashMap, Map, TreeMap}
import scala.jdk.CollectionConverters.*

class GenEntailVerifier(val program: Program, order: String) {

  private var counter: Int = 0
  private val termManager: TermManager = new TermManager()
  private val sort: Sort = termManager.mkFiniteFieldSort(order, 10)

  private def genFreshString(): Str = {
    counter += 1
    Str("$" + counter)
  }

  private def genFreshCID(): Num = {
    counter += 1
    Num(-counter)
  }

  def genFreshValue(`type`: Type): Expr = `type` match
    case _: StringType      => genFreshString()
    case _: PartyIndexType  => genFreshCID()
    case t: RecordType =>
      val map: TreeMap[Identifier, Expr] = new TreeMap()
      t.elements.asScala.foreach { case (k, v) =>
        map.put(k, genFreshValue(v))
      }
      FieldExpr(map)
    case other =>
      throw new RuntimeException("Unsupported type " + other)

  /**
   * Generalized rule for entailment: verify if precondition entails postcondition for all inputs
   *
   * @param typings list of typed identifiers
   * @param e1      abstract precondition
   * @param e2      abstract postcondition
   * @return true/false
   */
  def genEntails(typings: java.util.List[TypedIdentifier], e1: Constraint, e2: Constraint): Boolean = {
    val bindingList: Map[Identifier, Expr] = new HashMap()

    // generate fresh values for variables
    typings.asScala.foreach { typing =>
      bindingList.put(typing.y, genFreshValue(typing.t))
    }

    // evaluate abstract constraints into ground constraints
    val evaluator = new ConstraintEvaluator(program)
    evaluator.binding_list.add(bindingList)
    val groundE1: Constraint = evaluator.evalConstraint(e1)
    val groundE2: Constraint = evaluator.evalConstraint(e2)
    evaluator.binding_list.removeLast()

    // check if ground constraints are fully instantiated
    if ConstraintChecker.checkConstraint(groundE1) && ConstraintChecker.checkConstraint(groundE2) then
      val termFactory = new TermFactory(termManager, sort)
      val groundE1Term = termFactory.constraintToTerm(groundE1)
      val groundE2Term = termFactory.constraintToTerm(groundE2)
      val verifier = new Verifier(termFactory)
      verifier.entails(groundE1Term, groundE2Term)
    else
      throw new RuntimeException("constraints with fresh values are not ground!")
  }
}
